specification/0041-hashed-mappings-specification.md

```diff
@@ -0,0 +1,143 @@
+| ------------------------------------- | ----------------------------- | ------------- |
+| `org.example.ClassA`                  | `ClassA`                      | `C_fshauzuk`  |
+| `org.example.ClassA.InnerClass`       | `ClassA$InnerClass`           | `C_zldtvwcl`  |
+| `org.example.ClassA.fieldA`           | `f;ClassA.fieldA;`            | `f_xskwwtwz`  |
+| `org.example.ClassA.methodA()`        | `m;ClassA.methodA;`           | `m_onqxeppm`  |
+| `org.example.ClassB`                  | `ClassB`                      | `C_irypywpy`  |
+| `org.example.ClassB.InnerClass`       | `ClassB$InnerClass`           | `C_znuawieq`  |
+| `org.example.ClassB.methodA()`        | `m;ClassA.methodA;`           | `m_onqxeppm`  |
+| `org.example.ClassB.methodB()`        | `m;InterfaceA.methodB;()V`    | `m_reinsnrb`  |
+| `org.example.InterfaceA`              | `InterfaceA`                  | `C_wtvyrzif`  |
+| `org.example.InterfaceA.methodA()`    | `m;ClassA.methodA;`           | `m_onqxeppm`  |
+| `org.example.InterfaceA.methodB()`    | `m;InterfaceA.methodB;()V`    | `m_reinsnrb`  |
+| `org.example.InterfaceA.methodB(int)` | `m;InterfaceA.methodB;(I)V`   | `m_mkznfdir`  |
+| `org.example.packageA.ClassC`         | `org/example/packageA/ClassC` | `C_hehlvtlo`  |
+| `org.example.packageB.ClassC`         | `org/example/packageB/ClassC` | `C_rvxxwsev`  |
+
+## Drawbacks
+
+- This version of hashed is potentially still less stable than intermediary, although this has not been quantified.
+  - An "ideal" intermediary would be perfectly stable.
+- Hashed base-26 names might be less readable than other name formats.
+  - This is unfortunately hard to quantify. Base-10 hashing would be an alternative, although with longer hashes.
+- A specification of the hashed mappings might make the format inflexible.
+  - Changes in hashed already come with great effects, so any such change should probably go through the RFC process anyway.
+
+## Rationale and Alternatives
+
+- This version of hashed is most likely more stable than Mojmap.
+  - Testing has shown that around 50% of class renames in Mojmap are only repackages,
+    which are avoided thanks to package stripping.
+- Hashed mappings can be fully automated, saving us the work of matching intermediary manually.
+  - We've already evaluated that a fully automated system can update hashed within less than 10 minutes of a new release.
+- When working with layered mappings (e. g. quilt mappings on top of hashed) it is preferable
+  to be able to clearly distinguish between mapped and unmapped code.
+  - Encountering Mojmap names when working with QM could be highly confusing.
+- Unique hashed names provide several benefits:
+  - Easier deobfuscation of stacktraces and other errors (especially when using bots for manual translation).
+  - Allow source remapping from hashed to any other mapping without needing context (no classpath or even full class files needed).
+  - Un-hides fields.
```

Comment:
What is meant here by 'unhide fields'?

Reply:
That is a really good question, and I'm not sure if I can answer that, but I'll try:

Hiding a field happens when a sub-class declares a field with the same name and descriptor as the super class.
This means that the sub class can't access the super class field, unless it uses the super keyword.
Since hashed makes every field name unique, it un-hides all fields.

Now the real question is, does it matter? I think yes:

Imagine a `class B extends A`, both have a field called `foo` in hashed.
Now quilt mappings decides to give the fields different names, `B.fooB` and `A.fooA`.
Now imagine writing some code that accesses both `A.fooA` and `A.fooB`.
After remapping to hashed, the code would read `A.foo` and `A.foo`, which is not the intended behaviour.

This might not matter, depending on if the compiler can ever produce this example,

Reply:
Ah, I found the relevant reference in JLS 16 (also applies in JLS 17):
> If a class declares a field with a certain name, then the declaration of that field is said to _hide_ any and all accessible declarations of fields with the same name in superclasses, and superinterfaces of the class.
> ~ _[The Java™ Language Specification, Java SE 16 Edition, &sect; 8.3. Field Declarations ](https://docs.oracle.com/javase/specs/jls/se16/html/jls-8.html#jls-8.3)_

---

> Hiding a field happens when a sub-class declares a field with the same name and descriptor as the super class.

Almost accurate, but according to the same JLS section:
> If a field declaration hides the declaration of another field, the two fields need not have the same type. 
> ~ _[The Java™ Language Specification, Java SE 16 Edition, &sect; 8.3. Field Declarations ](https://docs.oracle.com/javase/specs/jls/se16/html/jls-8.html#jls-8.3)_

Reply:
Yes. the "hiding" terminology is from JLS, but in the JVM they would also have to share the same descriptor.